### PR TITLE
8264805: Remove the experimental Ahead-of-Time Compiler

### PR DESCRIPTION
**Pull request by vnkozlov**
As part of [JEP 410](http://openjdk.java.net/jeps/410) remove code related to Ahead-of-Time Compiler from JDK: 

- `jdk.aot` module — the `jaotc` tool 
- `src/hotspot/share/aot` — loads AoT compiled code into VM for execution 
- related HotSpot code guarded by `#if INCLUDE_AOT` 

Additionally, remove tests as well as code in makefiles related to AoT compilation.

Tested hs-tier1-4

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264805](https://bugs.openjdk.java.net/browse/JDK-8264805): Remove the experimental Ahead-of-Time Compiler

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Igor Ignatyev](https://openjdk.java.net/census#iignatyev) (@iignatev - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 6cce0f6cad6fe6ab8815863acb979b4a4c1ba6c9
 * [Igor Veresov](https://openjdk.java.net/census#iveresov) (@veresov - **Reviewer**) ⚠️ Review applies to 71a166c14d056d0837a7d86c27e1708651bd46f3

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3398/head:pull/3398` \
`$ git checkout pull/3398`

Update a local copy of the PR: \
`$ git checkout pull/3398` \
`$ git pull https://git.openjdk.java.net/jdk pull/3398/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3398`

View PR using the GUI difftool: \
`$ git pr show -t 3398`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3398.diff">https://git.openjdk.java.net/jdk/pull/3398.diff</a>

</details>

### CODE REVIEW

**Comment by coleenp** [comment]
void CodeCache::mark_for_evol_deoptimization(InstanceKlass* dependee) {
This function, its caller and the function in jvmtiRedefineClasses.cpp that calls it can be deleted also, but you can file a separate RFE for that if you want.

**Review by erikj79 — APPROVED** [review]
Build changes look good.

**Review by stefank — APPROVED** [review]
GC changes look good.

**Comment by vnkozlov** [comment]
> void CodeCache::mark_for_evol_deoptimization(InstanceKlass* dependee) {
> This function, its caller and the function in jvmtiRedefineClasses.cpp that calls it can be deleted also, but you can file a separate RFE for that if you want.

Thank you, Coleen, for review. I will wait other comments and will remove this code.

**Comment by vnkozlov** [comment]
Thank you, Erik and Stefan.

**Review by iignatev — APPROVED** [review]
Test changes look good to me.

**Comment by coleenp** [comment]
There's a comment above
void VM_RedefineClasses::mark_dependent_code(InstanceKlass* ik) {
that should be rewritten, so I think we should just file an RFE to remove it afterwards.

**Comment by vnkozlov** [comment]
> There's a comment above
> void VM_RedefineClasses::mark_dependent_code(InstanceKlass* ik) {
> that should be rewritten, so I think we should just file an RFE to remove it afterwards.

https://bugs.openjdk.java.net/browse/JDK-8264941

**Review by dholmes-ora — APPROVED** [review]
Hi Vladimir,

This looks good to me - I went through all the files.

It was nice to see how nicely compartmentalized the AOT feature was - that made checking the changes quite simple. The one exception was the fingerprinting code, which for some reason was AOT-only but not marked that way, so I'm not sure what the story is there. ??

I was also surprised to see some of the flags were not marked experimental, so there will need to a CSR request to remove those without going through the normal deprecation process.

Thanks,
David

**Review by theRealAph — APPROVED** [review]
AArch64n looks fine.

File: src/hotspot/cpu/x86/globalDefinitions_x86.hpp [hunk 1]
```diff
@@ -70,7 +70,7 @@ const bool CCallingConventionRequiresIntsAsLongs = false;
 #endif
 
 #if INCLUDE_JVMCI
-#define COMPRESSED_CLASS_POINTERS_DEPENDS_ON_COMPRESSED_OOPS (EnableJVMCI || UseAOT)
+#define COMPRESSED_CLASS_POINTERS_DEPENDS_ON_COMPRESSED_OOPS (EnableJVMCI)
```
Comment (shipilev) [inline src/hotspot/cpu/x86/globalDefinitions_x86.hpp:73]: Minor nit: can probably drop parentheses here.

File: src/hotspot/share/code/compiledIC.cpp [hunk 2]
```diff
@@ -734,10 +709,10 @@ void CompiledDirectStaticCall::print() {
     tty->print("clean");
   } else if (is_call_to_compiled()) {
     tty->print("compiled");
-  } else if (is_call_to_far()) {
-    tty->print("far");
   } else if (is_call_to_interpreted()) {
     tty->print("interpreted");
+  } else {
+    tty->print("unknown");
```
Comment (shipilev) [inline src/hotspot/share/code/compiledIC.cpp:715]: We can probably split this cleanup into the minor issue, your call. The benefit of separate issue: backportability.

**Review by shipilev — APPROVED** [review]
Shenandoah parts look good. I have a few minor comments after reading the rest.

File: src/hotspot/share/oops/instanceKlass.hpp [hunk 3]
```diff
@@ -255,8 +254,7 @@ class InstanceKlass: public Klass {
     _misc_has_nonstatic_concrete_methods      = 1 << 5,  // class/superclass/implemented interfaces has non-static, concrete methods
     _misc_declares_nonstatic_concrete_methods = 1 << 6,  // directly declares non-static, concrete methods
     _misc_has_been_redefined                  = 1 << 7,  // class has been redefined
-    _misc_has_passed_fingerprint_check        = 1 << 8,  // when this class was loaded, the fingerprint computed from its
-                                                         // code source was found to be matching the value recorded by AOT.
+    _unused                                   = 1 << 8,  //
```
Comment (veresov) [inline src/hotspot/share/oops/instanceKlass.hpp:257]: Any particular reason we don't want to remove this gap?

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp [hunk 4]
```diff
@@ -1331,12 +1184,6 @@ void CodeInstaller::site_DataPatch(CodeBuffer& buffer, jint pc_offset, JVMCIObje
       if (!_immutable_pic_compilation) {
```
Comment (veresov) [inline src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:1184]: All _immutable_pic_compilation mentions can be removed as well. It is true only for AOT compiles produced by Graal. It's never going to be used without AOT.

File: src/hotspot/share/oops/methodCounters.cpp [hunk 5]
```diff
@@ -79,9 +76,6 @@ void MethodCounters::clear_counters() {
 
 void MethodCounters::metaspace_pointers_do(MetaspaceClosure* it) {
```
Comment (iklam) [inline src/hotspot/share/oops/methodCounters.cpp:77]: MethodCounters::metaspace_pointers_do can be removed altogether (also need to remove the declaration in methodCounter.hpp).

**Review by iklam — APPROVED** [review]
LGTM. Just a small nit.

**Comment by vnkozlov** [comment]
> Hi Vladimir,
> 
> This looks good to me - I went through all the files.
> 
> It was nice to see how nicely compartmentalized the AOT feature was - that made checking the changes quite simple. The one exception was the fingerprinting code, which for some reason was AOT-only but not marked that way, so I'm not sure what the story is there. ??
> 
> I was also surprised to see some of the flags were not marked experimental, so there will need to a CSR request to remove those without going through the normal deprecation process.
> 
> Thanks,
> David

Thank you, David.
We thought that we could use fingerprint code for other cases that is why we did not put it under `#if INCLUDE_AOT`.
I filed CSR for AOT product flags removal: https://bugs.openjdk.java.net/browse/JDK-8265000

**Comment by vnkozlov** [comment]
Thank you, Igor Ignatyev, Igor Veresov, Ioi, Aleksey and Andrew for reviews.
I will update changes based on your comments.

File: src/hotspot/cpu/x86/compiledIC_x86.cpp [hunk 6]
```diff
@@ -193,11 +131,11 @@ void CompiledDirectStaticCall::verify() {
 
 #ifdef ASSERT
   CodeBlob *cb = CodeCache::find_blob_unsafe((address) _call);
-  assert(cb && !cb->is_aot(), "CompiledDirectStaticCall cannot be used on AOTCompiledMethod");
+  assert(cb, "sanity");
```
[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/cpu/x86/compiledIC_x86.cpp:134]: done

File: src/hotspot/share/memory/heap.hpp [hunk 7]
```diff
@@ -171,15 +171,7 @@ class CodeHeap : public CHeapObj<mtCode> {
...
-    // AOT CodeBlobs (i.e. AOTCompiledMethod) objects aren't allocated in the AOTCodeHeap but on the C-Heap.
-    // Only the code they are pointing to is located in the AOTCodeHeap. All other CodeBlobs are allocated
-    // directly in their corresponding CodeHeap with their code appended to the actual C++ object.
-    // So all CodeBlobs except AOTCompiledMethod are continuous in memory with their data and code while
-    // AOTCompiledMethod and their code/data is distributed in the C-Heap. This means we can use the
-    // address of a CodeBlob object in order to locate it in its heap while we have to use the address
-    // of the actual code an AOTCompiledMethod object is pointing to in order to locate it.
-    // Notice that for an ordinary CodeBlob with code size zero, code_begin() may point beyond the object!
-    const void* start = AOT_ONLY( (code_blob_type() == CodeBlobType::AOT) ? blob->code_begin() : ) (void*)blob;
+    const void* start = (void*)blob;
```
[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/share/memory/heap.hpp:174]: Done:
```
   bool contains_blob(const CodeBlob* blob) const {
-    const void* start = (void*)blob;
-    return contains(start);
+    return contains((void*)blob);
   }
```

File: src/hotspot/cpu/x86/globalDefinitions_x86.hpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/cpu/x86/globalDefinitions_x86.hpp:73]: done

File: src/hotspot/share/code/compiledIC.cpp (hunk 2 above)
Comment (vnkozlov) [reply src/hotspot/share/code/compiledIC.cpp:715]: Reverted and filed https://bugs.openjdk.java.net/browse/JDK-8265013

File: src/hotspot/share/oops/instanceKlass.hpp (hunk 3 above)
Comment (vnkozlov) [reply src/hotspot/share/oops/instanceKlass.hpp:257]: Less changes. We don't get any benefits from removing it. It is opposite - if we need a new value we will use it without changing following values again.

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp (hunk 4 above)
Comment (vnkozlov) [reply src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:1184]: Done. I removed _immutable_pic_compilation in JVMCI in Hotspot.

File: src/hotspot/share/oops/methodCounters.cpp (hunk 5 above)
Comment (vnkozlov) [reply src/hotspot/share/oops/methodCounters.cpp:77]: Done.

**Review by mlchung — APPROVED** [review]
I reviewed the module-loader-map and non-hotspot non-AOT tests.

**Review by veresov — APPROVED** [review]

**Comment by dean-long** [comment]
@iklam
I thought the fingerprint code was also used by CDS.

**Comment by iklam** [comment]
> @iklam
> I thought the fingerprint code was also used by CDS.

CDS actually uses a different mechanism (CRC of the classfile bytes) to validate that a class has not changed between archive dumping time and runtime. See

https://github.com/openjdk/jdk/blob/5784f6b7f74d0b8081ac107fea172539d57d6020/src/hotspot/share/classfile/systemDictionaryShared.cpp#L1126-L1130

**Pull request closed without merging**